src/algebra/squarefree.lean

```diff
@@ -83,6 +86,33 @@ begin
   norm_cast,
 end
 
+end comm_monoid
+
+section cancel_comm_monoid_with_zero
+
+variables [cancel_comm_monoid_with_zero R] [decidable_rel (has_dvd.dvd : R → R → Prop)]
+  [wf_dvd_monoid R]
+
+lemma finite_prime_left {a b : R} (ha : prime a) (hb : b ≠ 0) :
+  multiplicity.finite a b :=
+begin
+  revert hb,
+  refine wf_dvd_monoid.induction_on_irreducible b _ _ _,
```

Comment:
you can also use λs in the `refine` to avoid having to `intro`, but it's no big deal. I'd write this line, for example, as `refine wf_dvd_monoid.induction_on_irreducible b _ (λ u hu hu', _) (λ b p hb hp ih hpb, _)`, and if there was a bit more space I'd even replace the first dash with `(by contradiction)`. just a stylistic thing though!